<h1> Analiza transportnih podatkov </h1>

<h2> Pregled in urejanje podatkov</h2>

In [219]:
# naložimo Pandas 
import pandas as pd
# import os.path -> Včasih problem, za različne operacijske sisteme

# izberemo interaktivni "notebook" stil risanja
# %matplotlib notebook

#Uvoz podatkov-----------
pathData = "data_set - transported.csv"
data = pd.read_csv(pathData)
#------------------------

# Izpisovanje -----------
pd.options.display.max_rows = 20
pd.options.display.max_columns = len(data.columns)
#------------------------


# Številske stolpce Str -> Int
numberColumns = ['Deliveries', 'Attachments', 'Unique destinations']

for i in numberColumns:
    data[i] = data[i].astype(int)

# Weight stolpec Str -> float

data['Weight'] = data['Weight'].apply(lambda x: float(x.replace(',','')))

#------------------------

# data.columns = data.columns.str.replace(' ', '')


In [220]:
data

ID ETA status ETA type Difference with ETA  Deliveries  Attachments  \
0      1       None      NaN                               1            0   
1      2       None      NaN                               1            0   
2      3       None      NaN                               1            0   
3      4       None      NaN                               1            0   
4      5       None      NaN                               1            0   
..   ...        ...      ...                 ...         ...          ...   
386  387       None      NaN                               1            0   
387  388       None      NaN                               1            0   
388  389       None      NaN                               2            0   
389  390       None      NaN                               1            0   
390  391       None      NaN                               1            0   

    Loading date Loading period Unloading date Unloading period  ZIP - Origin  \
0     2022-09-01  00:00 - 23:59     2022-09-05    08:00 - 17:00         52333   
1     2022-09-01  00:00 - 23:59     2022-09-05    08:00 - 17:00         52333   
2     2022-09-01  00:00 - 23:59     2022-09-05    07:30 - 12:00         52333   
3     2022-09-01  00:00 - 23:59     2022-09-02    08:00 - 16:00         52333   
4     2022-09-01  00:00 - 23:59     2022-09-02    08:00 - 16:00         52333   
..           ...            ...            ...              ...           ...   
386   2022-08-26  00:00 - 00:00     2022-08-28    00:00 - 00:00          3273   
387   2022-08-25  00:00 - 00:00     2022-08-25    00:00 - 00:00          3273   
388   2022-08-24  00:00 - 00:00     2022-08-26    00:00 - 00:00          3273   
389   2022-08-09  00:00 - 00:00     2022-08-10    00:00 - 00:00          3273   
390   2022-08-03  00:00 - 00:00     2022-08-04    00:00 - 00:00          3273   

    City - Origin  Region - Origin Country - Origin  Unique destinations  \
0        Potpican              NaN               HR                    1   
1        Potpican              NaN               HR                    1   
2        Potpican              NaN               HR                    1   
3        Potpican              NaN               HR                    1   
4        Potpican              NaN               HR                    1   
..            ...              ...              ...                  ...   
386    HALMAJUGRA              NaN               HU                    1   
387    HALMAJUGRA              NaN               HU                    1   
388    HALMAJUGRA              NaN               HU                    2   
389    HALMAJUGRA              NaN               HU                    1   
390    HALMAJUGRA              NaN               HU                    1   

     ZIP - Destination           City - Destination  Region - Destination  \
0                15068            Pozzolo Formigaro                   NaN   
1                64010                     Ancarano                   NaN   
2                67100               Poggio Picenze                   NaN   
3                37060  RONCANOVA DI GAZZO VERONESE                   NaN   
4                20865                       USMATE                   NaN   
..                 ...                          ...                   ...   
386              31216                    Antunovac                   NaN   
387              31216                    Antunovac                   NaN   
388              32270                      Zupanja                   NaN   
389              34000                       Požega                   NaN   
390              34000                       Pozega                   NaN   

    Country - Destination  Quote request type  Quote request  \
0                      IT                 NaN            0.0   
1                      IT                 NaN            0.0   
2                      IT                 NaN            0.0   
3             

<h3> Različne vrednosti </h3>

<p>Pogledamo koliko je različnih vrednosti v posameznih stolpcih. </p>

In [221]:
#Koliko je različnih vrednosti za posamezen stolpec?

uniqueDataDic = {}
uniqueDataDicCount = {}

for type in data.columns[1:]:
    x = data[type].unique().tolist()
    uniqueDataDic[type] = x
    uniqueDataDicCount[type] = [len(x)]

uniqueDataCount = pd.DataFrame.from_dict(uniqueDataDicCount)

uniqueDataCount

ETA status  ETA type  Difference with ETA  Deliveries  Attachments  \
0           3         2                    6           3            3   

   Loading date  Loading period  Unloading date  Unloading period  \
0            59              19              64                11   

   ZIP - Origin  City - Origin  Region - Origin  Country - Origin  \
0             3              3                1                 3   

   Unique destinations  ZIP - Destination  City - Destination  \
0                    3                139                 145   

   Region - Destination  Country - Destination  Quote request type  \
0                     1                      7                   1   

   Quote request  Quote request currency  Quote request comment  \
0              1                       1                      1   

   Vehicle (Requirement)  Weight  Weight unit  
0                      5     165            1

In [222]:
#poglejmo neočitne vnose po stolpcih

notObviousData = ['ETA status','ETA type','Difference with ETA','Deliveries','Attachments','City - Origin','Country - Origin','Country - Destination','Unique destinations','Vehicle (Requirement)']

#City - Destination -> preveč različnih

for type in notObviousData:
    print(f"{type} : {uniqueDataDic[type]}")

ETA status : ['None', 'On time', 'Delayed']
ETA type : [nan, 'Destination']
Difference with ETA : [' ', '-01:00:43', '-00:09:14', '-00:06:20', '-00:15:03', '00:00:58']
Deliveries : [1, 3, 2]
Attachments : [0, 1, 2]
City - Origin : ['Potpican', 'Caparroso', 'HALMAJUGRA']
Country - Origin : ['HR', 'ES', 'HU']
Country - Destination : ['IT', 'BA', 'HU', 'HR', 'SI', 'RS', 'AT']
Unique destinations : [1, 3, 2]
Vehicle (Requirement) : ['Truck 100 m3 NO', 'Truck 90 m3', 'Megatrailer 95 m3', 'Truck 120 m3 NO', 'HU 24T one-way max 2 stops (5000)']


<p> Že tukaj lahko vidimo neke osnovne podatke o delovanju podjetja.</p>
<ul>
<li>Podjetje opravlja na prevoz maksimalno 3 dostave.</li>
<li>Prevoz se začne v treh mestih v treh različnih državah: Hr - Potpican, Es - Caparroso, Hu - Halmajugra.</li>
<li>Prevaža se v 7 različnih držav in 145 različnih mest.</li>
<li>Za različne dostave potrtebujejo 5 tipov tovornjakov.</li>
</ul>

<h3>Datum</h3>

<p>Uredimo datum. Dani zapis %Y-%m-%d ločimo, dodamo stolpce za mesec in dan.</p>

In [223]:
# Iz datuma dodamo določimo mesec, dan


#Loading date -> v datetime zapis
data['Loading date'] = pd.to_datetime(data['Loading date'],format='%Y-%m-%d')
data['Unloading date'] = pd.to_datetime(data['Unloading date'],format='%Y-%m-%d')

#Dodajanje stolpcev day, month
data['Loading month'] = data['Loading date'].dt.month
data['Loading day'] = data['Loading date'].dt.day

data['Unloading month'] = data['Unloading date'].dt.month
data['Unloading day'] = data['Unloading date'].dt.day


#Časovno trajanje poti

#Pazimo, če se začne en mesec in konča drugi mesec
#data['Delivery duration'] = (data["Unloading day"] - data['Loading day']) if (data["Unloading month"] == data['Loading month']) else None

data['Duration (days)'] = (data['Unloading date'] - data['Loading date']).dt.days

data

ID ETA status ETA type Difference with ETA  Deliveries  Attachments  \
0      1       None      NaN                               1            0   
1      2       None      NaN                               1            0   
2      3       None      NaN                               1            0   
3      4       None      NaN                               1            0   
4      5       None      NaN                               1            0   
..   ...        ...      ...                 ...         ...          ...   
386  387       None      NaN                               1            0   
387  388       None      NaN                               1            0   
388  389       None      NaN                               2            0   
389  390       None      NaN                               1            0   
390  391       None      NaN                               1            0   

    Loading date Loading period Unloading date Unloading period  ZIP - Origin  \
0     2022-09-01  00:00 - 23:59     2022-09-05    08:00 - 17:00         52333   
1     2022-09-01  00:00 - 23:59     2022-09-05    08:00 - 17:00         52333   
2     2022-09-01  00:00 - 23:59     2022-09-05    07:30 - 12:00         52333   
3     2022-09-01  00:00 - 23:59     2022-09-02    08:00 - 16:00         52333   
4     2022-09-01  00:00 - 23:59     2022-09-02    08:00 - 16:00         52333   
..           ...            ...            ...              ...           ...   
386   2022-08-26  00:00 - 00:00     2022-08-28    00:00 - 00:00          3273   
387   2022-08-25  00:00 - 00:00     2022-08-25    00:00 - 00:00          3273   
388   2022-08-24  00:00 - 00:00     2022-08-26    00:00 - 00:00          3273   
389   2022-08-09  00:00 - 00:00     2022-08-10    00:00 - 00:00          3273   
390   2022-08-03  00:00 - 00:00     2022-08-04    00:00 - 00:00          3273   

    City - Origin  Region - Origin  ... Country - Destination  \
0        Potpican              NaN  ...                    IT   
1        Potpican              NaN  ...                    IT   
2        Potpican              NaN  ...                    IT   
3        Potpican              NaN  ...                    IT   
4        Potpican              NaN  ...                    IT   
..            ...              ...  ...                   ...   
386    HALMAJUGRA              NaN  ...                    HR   
387    HALMAJUGRA              NaN  ...                    HR   
388    HALMAJUGRA              NaN  ...                    HR   
389    HALMAJUGRA              NaN  ...                    HR   
390    HALMAJUGRA              NaN  ...                    HR   

     Quote request type  Quote request Quote request currency  \
0                   NaN            0.0                    NaN   
1                   NaN            0.0                    NaN   
2                   NaN            0.0                    NaN   
3                   NaN            0.0                    NaN   
4                   NaN            0.0                    NaN   
..                  ...            ...                    ...   
386                 NaN            0.0                    NaN   
387                 NaN            0.0                    NaN   
388                 NaN            0.0                    NaN   
389                 NaN            0.0                    NaN   
390                 NaN            0.0                    NaN   

     Quote request comment              Vehicle (Requirement)    Weight  \
0                      NaN                    Truck 100 m3 NO   7707.85   
1                      NaN                    Truck 100 m3 NO   8843.54   
2                      NaN                    Truck 100 m3 NO   6134.11   
3                      NaN                    Truck 100 m3 NO   5742.88   
4                      NaN                    Truck 100 m3 NO   9327.65   
..                     ...                                ...       ...   
386                  

In [224]:
#Za katere mesece so podatki

type = 'Loading month'
print( f"Loading month : {data[type].unique().tolist()}")

Loading month : [9, 8, 7, 6, 5]


<h3> Odstarnjevanje nepotrebnih podatkov</h3>
 
<p>Poglejmo le uporabne podatke. Odsatrnimo notUseful stolpce. Glej kodo.</p>

<p>Spodaj so podani urejeni podatki.</p>

In [228]:
notUseful = ['Loading date', 'Unloading date', 'Loading period', 'Unloading period', 'ZIP - Origin', 'Region - Origin', 'ZIP - Destination', 'Region - Destination', 'Quote request type', 'Quote request', 'Quote request currency', 'Quote request comment']

useful = [i for i in data.columns.to_list() if i not in notUseful]

sortedData = data.loc[:, useful]

sortedData

ID ETA status ETA type Difference with ETA  Deliveries  Attachments  \
0      1       None      NaN                               1            0   
1      2       None      NaN                               1            0   
2      3       None      NaN                               1            0   
3      4       None      NaN                               1            0   
4      5       None      NaN                               1            0   
..   ...        ...      ...                 ...         ...          ...   
386  387       None      NaN                               1            0   
387  388       None      NaN                               1            0   
388  389       None      NaN                               2            0   
389  390       None      NaN                               1            0   
390  391       None      NaN                               1            0   

    City - Origin Country - Origin  Unique destinations  \
0        Potpican               HR                    1   
1        Potpican               HR                    1   
2        Potpican               HR                    1   
3        Potpican               HR                    1   
4        Potpican               HR                    1   
..            ...              ...                  ...   
386    HALMAJUGRA               HU                    1   
387    HALMAJUGRA               HU                    1   
388    HALMAJUGRA               HU                    2   
389    HALMAJUGRA               HU                    1   
390    HALMAJUGRA               HU                    1   

              City - Destination Country - Destination  \
0              Pozzolo Formigaro                    IT   
1                       Ancarano                    IT   
2                 Poggio Picenze                    IT   
3    RONCANOVA DI GAZZO VERONESE                    IT   
4                         USMATE                    IT   
..                           ...                   ...   
386                    Antunovac                    HR   
387                    Antunovac                    HR   
388                      Zupanja                    HR   
389                       Požega                    HR   
390                       Pozega                    HR   

                 Vehicle (Requirement)    Weight Weight unit  Loading month  \
0                      Truck 100 m3 NO   7707.85          kg              9   
1                      Truck 100 m3 NO   8843.54          kg              9   
2                      Truck 100 m3 NO   6134.11          kg              9   
3                      Truck 100 m3 NO   5742.88          kg              9   
4                      Truck 100 m3 NO   9327.65          kg              9   
..                                 ...       ...         ...            ...   
386  HU 24T one-way max 2 stops (5000)  20197.80          kg              8   
387  HU 24T one-way max 2 stops (5000)  22157.80          kg              8   
388  HU 24T one-way max 2 stops (5000)  19628.00          kg              8   
389  HU 24T one-way max 2 stops (5000)  20785.80          kg              8   
390  HU 24T one-way max 2 stops (5000)  20785.80          kg              8   

     Loading day  Unloading month  Unloading day  Duration (days)  
0              1                9              5                4  
1              1                9              5                4  
2              1                9              5                4  
3              1                9              2                1  
4              1                9              2                1  
..           ...              ...            ...              ...  
386           26                8             28                2  
387           25                8             25                0  
388           24                8             26                2  
389            9                8             

<h2> Mesečna analiza</h2>

In [226]:
#Mesečna analiza

months = sortedData["Loading month"].unique().tolist()

dataMonthsDic = {'Months' : months}

sumWeight = []
sumDeliveries = []
countInput = []

for month in months:
    countInput.append(len(sortedData[data['Loading month'] == month]))
    sumWeight.append(sortedData.loc[data['Loading month'] == month, 'Weight'].sum())
    sumDeliveries.append(sortedData.loc[data['Loading month'] == month, 'Deliveries'].sum())

dataMonthsDic['All trips'] = countInput
dataMonthsDic['Sum Weight'] = sumWeight
dataMonthsDic['Sum Deliveries'] = sumDeliveries


dataMonths = pd.DataFrame.from_dict(dataMonthsDic)

dataMonths


#groupby
# data.groupby('Loading month')['Weight'].sum()

Months  All trips  Sum Weight  Sum Deliveries
0       9          5    37756.03               5
1       8        112   966294.18             121
2       7        137  1118766.29             143
3       6        135  1080662.47             143
4       5          2    20187.04               2

In [227]:
# International trips



In [ ]:
#Zamude 